### PR TITLE
on navigate (#332)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
```
from src.taipy.gui import Gui, State, Markdown

md_1 = """
This is page 1
<|{count}|number|label=counter|>
"""

md_2 = """
This is page 2
"""

md_3 = """
This is page 3
"""

md = """
# on_navigate test
<|navbar|lov={navlov}|>

<|content|>
"""

navlov = [
    ("/page1", "Page 1"),
    ("/page2", "Page 2"),
]

pages = {
    "/": Markdown(md),
    "page1": Markdown(md_1),
    "page2": Markdown(md_2),
    "page3": Markdown(md_3),
}

count = 0

def on_navigate(state: State, page: str) -> str:
    if page == "page2" and state.count == 1:
        return "page3"
    elif state.count == 2:
        return None
    return page

Gui(pages=pages).run(use_reloader=False)
```

### CODE REVIEW

File: src/taipy/gui/gui.py [hunk 1]
```diff
@@ -96,6 +96,9 @@ class Gui:
         on_init (Callable): The function that is called on the first connection of a new user.<br/>
             It defaults to the `on_init()` global function defined in the Python
             application.
+        on_navigate (Callable): The function that is called when a page is requested.<br/>
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py:100]: "It defaults to the `on_navigate()` global function defined in the Python  application."

I would add, just like for *on_init()*, that if there is no such function, nothing happens.

File: src/taipy/gui/gui.py [hunk 2]
```diff
@@ -1186,22 +1190,31 @@ def __init_route(self):
...
+        nav_page = page_name
+        if hasattr(self, "on_navigate") and callable(self.on_navigate):
+            try:
+                nav_page = self.on_navigate(self.__get_state(), page_name)
+                if not isinstance(nav_page, str):
+                    warnings.warn(f"on_navigate returned a invalid page name '{nav_page}'")
+                    nav_page = page_name
+            except Exception as e:
+                warnings.warn(f"Exception on on_navigate execution\n{e}")
+        page = next((page_i for page_i in self._config.pages if page_i._route == nav_page), None)
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py:1209]: If nav_page != page_name, you should I think call 'on_navigate' again...

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Small yet important change...

File: src/taipy/gui/gui.py (hunk 2 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui/gui.py:1209]: Why would I do this ?
if new_page != page the dev changed it in on_navigate ... If I call again on_navigate, how can the dev differentiate from a "normal" call ?

Comment (FabienLelaquais) [reply src/taipy/gui/gui.py:1209]: In my mind, it was to simplify the user's code, if there are several tests to perform before choosing a page name or another.
When I have a real use case, I'll let you know. Forget it for the time being.
Thx

**Review by FabienLelaquais — APPROVED** [review]
👍

**Review by dinhlongviolin1 — APPROVED** [review]
LGTM!

**Pull request merged**